fbgemm_gpu/src/jagged_tensor_ops.cu

```diff
@@ -143,6 +143,381 @@ __launch_bounds__(kMaxThreads) void jagged_dense_elementwise_dense_output_kernel
+  }
+};
+
+template <typename scalar_t>
+struct VecType64{};  // 64 bit load-store vector type
+
+template <>
+struct VecType64<c10::Half> {
+  typedef float2 TType; // Transaction Type
+  typedef struct __align__(8) {
+    __half a, b, c, d;
+  } half4;
+
+  union Data {
+    half4 val;
+    TType mask;
+  } data;
+
+  __device__ VecType64() {
+    data.mask = make_float2(0.0f, 0.0f);
+  }
+};
+
+template <>
+struct VecType64<c10::BFloat16> {
+  typedef float2 TType; // Transaction Type
+  typedef struct __align__(8) {
+    __nv_bfloat16 a, b, c, d;
+  } half4;
+
+  union Data {
+    half4 val;
+    TType mask;
+  } data;
+
+  __device__ VecType64() {
+    data.mask = make_float2(0.0f, 0.0f);
+  }
+};
+
+template <>
+struct VecType64<float> {
+  typedef float2 TType; // Transaction Type
+
+  union Data {
+    float2 val;
+    TType mask;
+  } data;
+
+  __device__ VecType64() {
+    data.mask = make_float2(0.0f, 0.0f);
+  }
+};
+
+template <>
+struct VecType64<double> {
+  typedef double TType; // Transaction Type
+
+  union Data {
+    double val;
+    TType mask;
+  } data;
+
+  __device__ VecType64() {
+    data.mask = 0.0;
+  }
+};
+
+template <>
+struct VecType64<int64_t> {
+  typedef int64_t TType; // Transaction Type
+
+  union Data {
+    int64_t val;
+    TType mask;
+  } data;
+
+  __device__ VecType64() {
+    data.mask = 0.0;
+  }
+};
+
+template <typename scalar_t>
+struct VecType32{};  // 64 bit load-store vector type
+
+template <>
+struct VecType32<c10::Half> {
+  typedef float TType; // Transaction Type
+
+  union Data {
+    __half2 val;
+    TType mask;
+  } data;
+
+  __device__ VecType32() {
+    data.mask = 0.0f;
+  }
+};
+
+template <>
+struct VecType32<c10::BFloat16> {
+  typedef float TType; // Transaction Type
+
+  union Data {
+    __nv_bfloat162 val;
+    TType mask;
+  } data;
+
+  __device__ VecType32() {
+    data.mask = 0.0f;
+  }
+};
+
+template <>
+struct VecType32<float> {
+  typedef float TType; // Transaction Type
+
+  union Data {
+    float val;
+    TType mask;
+  } data;
+
+  __device__ VecType32() {
+    data.mask = 0.0f;
+  }
+};
+
+template <>
+struct VecType32<double> {
+  typedef double TType; // Transaction Type
+};
+
+template <>
+struct VecType32<int64_t> {
+  typedef int64_t TType; // Transaction Type
+};
+
+template <typename index_t, typename scalar_t>
+__global__ void jagged_dense_elementwise_jagged_out_opt_kernel_(
+  index_t* dense_to_jagged_map,
+  scalar_t* dense,
+  scalar_t* jagged,
+  int32_t jagged_dim_max_length,
+  int32_t e_dim,
+  int32_t total_jagged_rows
+) {
+  // for some reason out rows in jagged can be larger than in dense
+  // for now assume same as elements in dense_to_jagged_map
+
+  typedef typename VecType128<scalar_t>::TType vec128;
+  typedef typename VecType64<scalar_t>::TType vec64;
+  typedef typename VecType32<scalar_t>::TType vec32; //  no-op for double
+
+  // 128 bit alignment check
+  if (((e_dim * sizeof(scalar_t)) << 3 ) % 128 == 0) {
```

Comment:
Instead of applying vectorization sort of all-or-nothing, can we do best effort like the first loop applied up to columns rounded down to 128-bit, the second loop applied to up to columns rounded down to 64-bit, and so on? In this way when e_dim is say 576-bit we can use 128-bit vectorization up to 512-bit and then 64-bit vectorization. Or this will regress perf?

Reply:
Performance will depend on the overhead of math instructions get added to do the alignment check by each thread of the warp every time it executes.  I think you can find both types of problem size where doing the incremental vector-bit decrement will help performance and other (perfectly aligned cases) where it will reduce it.
IMO ideal solution is to allocate contiguous tensors with appropriate pitch in each dimension that aligns the vectorized load/store. Not sure if pytorch does that already.

Reply:
So you're only checking alignment with respect to e_dim, not alignment of the actual pointers, right? If that's the case, can we just do the following? Or do we have to check alignment of "values" and "dense" pointers. 

```
    VecType128::TType * values128 = (VecType128::TType *)&(values[0]);
    VecType128::TType * dense128 = (VecType128::TType *)&(dense[0]);
    VecType64::TType * values64 = (VecType64::TType *)&(values[0]);
    VecType64::TType * dense64 = (VecType64::TType *)&(dense[0]);
    VecType32::TType * values32 = (VecType32::TType *)&(values[0]);
    VecType32::TType * dense32 = (VecType32::TType *)&(dense[0]);
    for(int real_row = values_row ; real_row < nnz ; real_row += blockDim.y * gridDim.y) {
        int dense_row = rows[real_row];
        int dense_col = cols[real_row];
        for(int tid = threadIdx.x ; tid < E/8 ; tid += blockDim.x) {
            values128[tid + real_row * (E/8)] = dense128[tid + dense_col * (E/8) + dense_row * L * (E/8)];
        }
        for(int tid = threadIdx.x + (E/8)*8 ; tid < E/4 ; tid += blockDim.x) {
            values64[tid + real_row * (E/4)] = dense64[tid + dense_col * (E/4) + dense_row * L * (E/4)];
        }
        for(int tid = threadIdx.x + (E/4)*4 ; tid < E/2 ; tid += blockDim.x) {
            values32[tid + real_row * (E/2)] = dense32[tid + dense_col * (E/2) + dense_row * L * (E/2)];
        }
        for(int tid = threadIdx.x + (E/2)*2 ; tid < E ; tid += blockDim.x) {
            values[tid + real_row * (E)] = dense[tid + dense_col * (E) + dense_row * L * (E)];
        }
    }
```

Reply:
I am doing a alignment check with `e_dim * sizeof(scalar_t)` because :
1. It will ensure jagged_out tensor is aligned
2. Because of broadcast and the only effective stride in dense tensor being same as e_dim, so my check ensures overall things will align for all the consequent pointer location that  kernel will try to read from. 

The alternate that you suggested will run into pointer alignment error. For eg: say output jagged is 130 bit wide (65 half's), so very first row will work fine with your logic, however the 2nd row will try to write first 8 half's with address that is not aligned for 128-bit operation.

> for(int tid = threadIdx.x ; tid < E/8 ; tid += blockDim.x) {
Note that in my code I use a warp to service two rows (block.x = 16) , I use `tid = threadIdx.x % blockDim.x`